[4.0] Fields subfields

continga opened the pull request:
### Summary of Changes

This is a new version of PR #22446 . We had a long discussion in #22446 about the feature of a subfields custom field, and after some feedback-cycles we came up with a good solution.

Sadly, that PR couldn't make it into the 3.x-branch, hence we need to rebase the whole PR and create a new one against 4.0-dev. This is the purpose of this PR. It will implement the functionality from #22446 into the 4.0-dev branch.

### Testing Instructions

See #22446 - basically what needs to be tested is the new "Subfields" custom field type, which can contain other custom field instances. It is basically a collection of other custom fields.

We need to check both the correct behaviour in the backend (entering custom field data, editing custom fields), as well as the default-display in the frontend (outputting all kind of different subfields+fields values, etc). See also further down at the docs-link.

### Expected result

See #22446

### Documentation Changes Required

Yes. I have started to work on it a bit, see https://docs.joomla.org/Custom_fields_type:_Subfields

I will update that docs-page over the coming weeks, it is currently not reflecting the very latest changes. But it is already giving a good quick overview about the feature, and possible cases to test.

## Additional information

I am very open for general code feedback, improvement ideas, test cases etc. I am not completely sure whether I followed all J4 coding-guidelines, so I would be glad to get some feedback and correct something - in case it's needed.

Generally, IMHO the feature is already very complex at the current state, so we should take distance from making it even more powerful. It makes more sense to get the current functionality merged, and then working on more advanced matters afterwards.

Also, it seems like choosing a category for a custom field currently seems broken in 4.0-dev generally, not only in this branch? Seems to be caused by an exception being thrown at https://github.com/joomla/joomla-cms/blob/e2abc0993baa4e5abb5a3166c56f3888c278c40c/administrator/components/com_fields/Model/FieldModel.php#L989 This needs to be investigated, but it ain't part of this PR. Just dropping this note here, because I think testers will stumble accross not being able to set a category for a custom field.

Summoning @regularlabs @AndySDH - you guys did a great job at testing the previous PR, maybe you could take a look at this one? :)

Commit:
Add the latest status of PR #22446 on top of the 4.0-dev branch, squashed.

This already contains adjustments for Joomla! 4, e.g. changed class files / paths, etc.
This does not yet contain any tests made or bugfixes against the 4.0 branch. This will
be done after this commit and possible bugfixes will be made in standalone commits.
So this basically only contains the code from #22446, adjusted to the 4.0 directory structure.

Commit:
Fixed immediate errors which preventing Joomla!4 to be installed after our last commit.

Commit:
Fix some syntax errors / deprecated stuff after switching to J4.

Commit:
Merge remote-tracking branch 'upstream/4.0-dev' into fields-subfields-4.0

Commit:
Adapt installation of plg_fields_subfields plugin to new autoincrement variant.

Commit:
Make phpcs happy.

Quy commented on administrator/components/com_fields/Helper/FieldsHelper.php:
```diff
@@ -626,6 +626,36 @@ public static function countItems(&$items)
+
+		if (!$fieldId)
+		{
+			return array();
```
```suggestion
			return [];
```

Quy commented on administrator/components/com_fields/tmpl/fields/default.php:
```diff
@@ -155,7 +155,12 @@
+	                                                    <?php $category_ids = FieldsHelper::getAssignedCategoriesIds($item->id); ?>
+	                                                    <?php if (in_array('-1', $category_ids)) : ?>
+		                                                    <?php echo JText::_('JNONE'); ?>
+	                                                    <?php else: ?>
```
```suggestion
	                                                    <?php else : ?>
```

Quy commented on administrator/components/com_fields/tmpl/fields/default.php:
```diff
@@ -155,7 +155,12 @@
                                                     <?php else: ?>
-                                                        <?php echo Text::_('JALL'); ?>
+	                                                    <?php $category_ids = FieldsHelper::getAssignedCategoriesIds($item->id); ?>
+	                                                    <?php if (in_array('-1', $category_ids)) : ?>
```
This can be changed to: 

```
<?php echo (in_array('-1', $category_ids)) ? Text::_('JNONE') : Text::_('JALL'); ?>
```

ghazal commented:
Tested but "Sub field *" dropdown doesn't show.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24711">issues.joomla.org/tracker/joomla-cms/24711</a>.</sub>

AndySDH commented:
Thanks a lot for your work!

I'll post things as I find them... :)

![image](https://user-images.githubusercontent.com/25923699/56655671-bedeaa00-6693-11e9-8873-858858a6787d.png)

A few notes about this...

1) Looks like Joomla 4 switched the true/false stuff to a toggle? See the "Required" option. 
So I'm guessing to keep it consistent, all the Yes/No stuff you have ("Repeatable" option and "Render values" options) should be changed to the same way of doing things? Not sure.

2) The tooltips for "Repeatable" and "Sub fields" is displayed under the option as well. Not sure why as it doesn't happen for the other options.
EDIT: Looks like option tooltips don't work at all in Joomla 4 :o

3) Looks like the option descriptions for "Sub field" and "Render values get repeated for every row instead of just being at the top for some reason. So this unnecessary and should be probably removed and only displayed at the top.

4) As it can obviously be seen, those long options descriptions break the layout and they don't wrap to new lines. But this is probably a general issue that has nothing to do with this PR. (such as the one you already mentioned for not being able to select a Category)

ghazal commented:
Sorry. After @AndySDH post, I understood my mistake.
Dropdown works allright.

Quy commented on plugins/fields/subfields/params/subfields.xml:
```diff
@@ -0,0 +1,60 @@
+				name="repeat"
+				type="radio"
+				label="PLG_FIELDS_SUBFIELDS_PARAMS_REPEAT_LABEL"
+				description="PLG_FIELDS_SUBFIELDS_PARAMS_REPEAT_DESC"
```
In J4, descriptions have been removed. The labels should be self-explanatory.

AndySDH replied:
That's terrible. Who made this idiotic decision?

HLeithner replied:
Only if the label is self explaining. It's not completely removed or?

AndySDH replied:
Now if you add a description it gets displayed UNDER the option and not as a tooltip anymore.
And now NONE of the core Joomla options labels make use of this description at all. So all core Joomla options tooltips are gone.

mbabker replied:
Please don't insult the work of others.  If you review the popovers in 3.x where these descriptions are heavily used, you will find in the majority of cases the descriptions were just a more verbose way of saying what the label already said.  If the description is actually adding value to a field, then it should be used, but the default stance with 4.0 is they are not used unless needed.

AndySDH replied:
I disagree. I'm looking at them right now. Many core tooltips were definitely adding value and explanation to the fields. And now none of the core tooltips are there at all.
Not to mention now extensions developers will have to do custom stuff to have the tooltpis back.
Sorry, but outright removing the feature is idiotic. It would have made a lot more sense to simply remove the verbose tooltips, if any, and leave the others alone.
But well, I guess it's decided anyway. I digress.

brianteeman replied:
1. It is not true that all descriptions have been removed
2. The fact that they are displayed underneath the field is currently a temporary measure until a decision is made on how to present them. There is at least one pr regarding this.

Why were they removed?
1. Because of the way bootstrap worked it wasn't possible to tell what had a tooltip and what did not. As a result everything was given a tooltip. The downside of that decision was that the vast majority of the descriptions were semantically identical to the label OR even worse were used instead of real documentation.
2. tooltips in their current form are completely inaccessible. They are only available to people using a mouse.

mbabker replied:
Support for the tooltips still exists in core, that hasn't been removed.

What was removed in core was things like the tooltip for an article's "Title" field which has a tooltip of either "Title" or "Title for the article." depending on which app's library language file is loaded.

The point is to address an accessibility issue (because the tooltips/popovers used fail in that department) and to remove redundant text from the UI.  The point is not to force everybody to not use the feature, it's to make people think about the text they're putting into the UI and either make sure things are clear or selectively use the description if it needs to be expanded on.

Just looking at this from a code perspective, the description for the "Render values" IMO makes sense to keep in.  The descriptions for these other fields fall into that bit of "is this really necessary?".  That's my opinion anyway.

AndySDH replied:
So support for tooltips is still here?

I understand the part about usability. But I don't think outright removing it from all core options is good. Some might have been verbose and unnededed, but some definitely not.

Just a couple of examples (out of many) of expanding descriptions that are necessary:
![image](https://user-images.githubusercontent.com/25923699/56659592-d1f67780-669d-11e9-8c0c-fbc844cdad39.png)

![image](https://user-images.githubusercontent.com/25923699/56659794-4c26fc00-669e-11e9-809a-305b821f9948.png)

Or descriptions for labels that were otherwise unclear:
![image](https://user-images.githubusercontent.com/25923699/56659680-0d914180-669e-11e9-9f08-dd44050ee4b9.png)

If you decided to go this way, at least a conscious review and selection should have been made on what labels were needing to keep a description (whether under it or wherever). But it looks like a "let's remove everything" approach was taken.

brianteeman replied:
NO it was a very conscious review that took me a very very long time. All three of your examples are in my opinion useless information that should not be there. What else could pagination results mean? 

Sorry you don't like it - but you can't please all the people all the time. Nor can you release software that is not accessible to everyone.

https://github.com/joomla/joomla-cms/pull/24711#discussion_r278093801

This is my final comment

AndySDH replied:
It's cool. The more I'm playing with it, the more I'm not a fan of the Joomla 4 admin UI and navigation/usability at all. So this tooltip thing is just one of the things that looks bad in my opinion. But indeed, as you said, you can't please all the people. So I guess I will have to just stay unpleased :)

AndySDH commented:
Some more stuff:


1) In article view, the layout doesn't seem to stack on mobile view
2) This issue is back:
- Issue with a calendar sub field: First filling (and saving) and then emptying a calendar sub field, makes it still store in the database value as "0000-00-00 00:00:00", and outputs it as "-0001-11-30". Instead if should obviously not be stored when empty.
3) Looks like there is space output around the subfield values, such as:
`<span class="field-value">          My Text     </span>`
4) So as we know, Repeatable Field needs to be gone in Joomla 4.0. So just a reminder that something will have to be done with this:
https://github.com/joomla/joomla-cms/issues/23659

Don't see anything else for now! :)

ghazal commented:
I have tested this item :white_check_mark: successfully on 174a97ad535e542c083b15ef8aecb9896ec35e11<br /><br />I really hope this will be implemented.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24711">issues.joomla.org/tracker/joomla-cms/24711</a>.</sub>

sanderpotjer commented:
@continga great work, thanks! Besides the feedback already shared by others i found an issue with the media field: The Media field is trowing JavaScript error if the article with field is not saved yet.  `Uncaught TypeError: this.querySelector(...).open is not a function`, so the media model is not opening

HLeithner commented:
@continga could you please resolve the conflicts?

Commit:
Merge remote-tracking branch 'upstream/4.0-dev' into fields-subfields-4.0

Commit:
Add formatting feedback from PR

Commit:
Remove unneeded descriptions for fields

Commit:
Change radio buttons to use class switcher

AndySDH commented:
Any update @continga?

AndySDH commented:
Any update @continga? Just making sure we don't forget this for Joomla 4! If @continga is not available, can someone else fix the conflicts?

Commit:
Merge remote-tracking branch 'upstream/4.0-dev' into fields-subfields-4.0

Commit:
Fix issue #24782:

This basically reverts the change made to this line in commit c778973, but that is a merge commit
and none of the merged parents have those change, and the commit does not give an explanation why
this change is needed. It is not needed at all from my point of view, so reverting it here.

This fixes #24782 and should not have any side effects, as the CategoryService is being instanciated
with a correct context (the component), so passing the $section is not needed. We want all categories.

Commit:
Merge remote-tracking branch 'origin/fix-24782' into fields-subfields-4.0

Commit:
Correct method signature to the parents'

Commit:
Fix issue #26133

This commit fixes issue #26133 The problem was caused by using the outdated Joomla.loadingLayer function,
which was replaced by a better approach. This commit integrates that approach.

Commit:
Merge remote-tracking branch 'origin/fix-26133' into fields-subfields-4.0

Commit:
Remove unneeded stuff, we removed the translation for this already, forgot to remove it here.

Commit:
Make phpcs happy.

continga commented:
Hi all, I just pushed an update for this which contains merging against 4.0-dev and some smaller bugfixes. Can y'all take a look again and report any last stuff? I am pretty sure we can somehow get this merged soon then yet.

PS: I don't know why the drone build is failing, looks like a CI failure and not something I have done wrong?

**Note for testing:** Currently, the selection of the sub fields stretches wide to the right, effectively overlaying with the content box to the right, hence not being able to use the add/remove/reorder buttons of the subform, see this screenshot:

![image](https://user-images.githubusercontent.com/2082625/64217380-76df5780-cebd-11e9-9c56-8bc305539cde.png)

This is being caused by the description of the top-row being way too long, and until we decide on a solution for this (maybe somehow just hide the description there, or implement this hotfix) there is a hotfix available to be able to use the page: Right-click on the "Render values" table header -> "Inspect element", remove the CSS rule "white-space: nowrap;", see screenshot:

![image](https://user-images.githubusercontent.com/2082625/64217426-a2fad880-cebd-11e9-95d9-432d8edc0d6b.png)

Then the layout looks better, not completely OK yet, but it is useable:

![image](https://user-images.githubusercontent.com/2082625/64217444-b5751200-cebd-11e9-95b9-38740f31759e.png)

Quy commented on administrator/components/com_fields/Field/SubfieldstypeField.php:
```diff
@@ -0,0 +1,133 @@
+
+use Joomla\CMS\Form\Field\ListField;
+use Joomla\Component\Fields\Administrator\Helper\FieldsHelper;
+use Joomla\CMS\HTML\HTMLHelper;
```
Sort in alpha order.

Quy commented on plugins/fields/subfields/tmpl/subfields.php:
```diff
@@ -0,0 +1,64 @@
+		}
+
+		// Generate the output for this sub field and row
+		$row_output[] = '<span class="field-entry' . ($class ? (' ' . $class) : '') . '">' . $content . '</span>';
```
```suggestion
		$row_output[] = '<span class="field-entry' . ($class ? ' ' . $class : '') . '">' . $content . '</span>';
```

continga replied:
I find this is really better with the paranthesis, much clearer, much clearer what's happening. Why do you want it to be removed?

Quy commented on plugins/fields/subfields/tmpl/subfields.php:
```diff
@@ -0,0 +1,64 @@
+}
+?>
+
+<?php if (trim($result) != '') : ?>
```
```suggestion
<?php if ($result != '') : ?>
```

continga replied:
Also here, why remove the `trim()`? Sure, **currently** this call is unneeded, since `$result` can either be completely empty, or always contains `<li>` or `</li>` at the end, but what if the code gets changed, someone puts spaces in between? I really want to make sure we only go into that IF, if `$result` not is blank.

AndySDH commented:
A suggestion. I think currently the variable names are misleading:

$row_output is actually the output of a subfield inside a row.

$result is actually the output of a row.

So I suggest:

$row_output -> $subfield_output
$result -> $row_output

Commit:
Merge remote-tracking branch 'upstream/4.0-dev' into fix-24782

Commit:
Fix getting categories for custom fields: Now trying component+section, and afterwards only component.

Commit:
Merge remote-tracking branch 'origin/fix-24782' into fields-subfields-4.0

Commit:
Try to make phpcs happy.

Commit:
Merge remote-tracking branch 'origin/fix-24782' into fields-subfields-4.0

continga commented:
@AndySDH I am not sure whether I understand your concern incorrectly, or whether you're wrong.

`$row_output` is an array containing the output of the subfields for one single row. So `$row_output` is the output for a single row. That is exactly what the variable name says. It is **not** the output of a single subfield inside a row - that is just what an array member of `$row_output` is.

Same for `$result`: It contains the overall result for all rows, **not** the output of a row. I also here think the name is correct.

Commit:
Alphanumerically sort the use statements.

AndySDH commented:
> @AndySDH I am not sure whether I understand your concern incorrectly, or whether you're wrong.
> 
> `$row_output` is an array containing the output of the subfields for one single row. So `$row_output` is the output for a single row. That is exactly what the variable name says. It is **not** the output of a single subfield inside a row - that is just what an array member of `$row_output` is.
> 
> Same for `$result`: It contains the overall result for all rows, **not** the output of a row. I also here think the name is correct.

Yup, you are right. I was thinking individual elements of the array instead of the whole array. It's actually good as it is :)

AndySDH commented:
I'm not able to test this in the latest Joomla Alha 11:

"The file marked for modification does not exist: build/media_source/system/js/fields/calendar.es5.js"

AndySDH commented:
So... I tested this with the latest J4 Alpha dev 12. The core backend template layout is still a complete mess so I'm not sure I'm able to judge this accurately when it comes to appearance.

Functionallity wise:

- Still an issue from before:  Issue with a calendar sub field: First filling (and saving) and then emptying a calendar sub field, makes it still store in the database value as "0000-00-00 00:00:00", and outputs it as "-0001-11-30". Instead if should obviously not be stored when empty.

- The "Repeatable" option currently can be changed after a subfield already exists. We figured it should not be possible to change it due to it messing up the values when changed.

I'll list what I can when it comes to appearance:

- As you already mentioned, it makes no sense for there being a nowrap CSS rule for th, it's only going to mess things up, so that should definitely be fixed in the Joomla template.

- Still an issue from before: When creating a subfield, the option descriptions for "Sub field" and "Render values get repeated for every row instead of just being at the top for some reason. So this unnecessary and should be probably removed and only displayed once at the top in the thead.

- Looks like now the Joomla template displays input fields side-by-side instead of stacking. In article edit form, the subfields type can't really coexist side by side with anything, as it's a table display in the first play,, so its control-group should always be forced to a new line and width 100%. See below the current situation:

![terrible](https://user-images.githubusercontent.com/25923699/64488403-df289300-d247-11e9-86e3-dcfad0d51423.JPG)

But again, the whole Joomla Backend template is pretty terrible right now so idk.

- Still an issue from before: In article edit form, the layout doesn't stack on mobile view

Commit:
Merge remote-tracking branch 'upstream/4.0-dev' into fields-subfields-4.0

continga commented:
Ok, how do we continue with this PR? I feel like the collective effort to get this merged is not very high, which I think is kind of sad after having received so much positive feedback ealier and after having spent so much work into this.

What we should do, from my point of view: We merge this into 4.0-dev now. The codebase for this PR is over one year old (see #22446), and no major issues have been found for a long time, and we had many different tests from different people in that timeframe. There are maybe 2 minor issues present at this time, and I will very gladly fix them after this got merged, because then I know my work is not being a pure waste of time. Additionally, we then get enough time to test this together with the whole community and react on possible problems.

The other possibility would be to just close this PR, as we would never get this merged otherwise. Joomla 4 Beta is about to sail away this month, and if we don't get this into beta we can just close it already.

What do we do?

HLeithner commented:
iirc https://github.com/joomla/joomla-cms/issues/23659 is missing for j4 incl. migration of old fields. ymmv

continga commented:
It is not "missing" in the sense of "it is required to be done" - it just **makes sense** to remove the repeatable type when this PR got merged, because the new type can do the same as the (then obsolete) repeatable type.

I will happily work in #23659 incl. a migration when we got this PR merged.

AndySDH commented:
Yes, this absolutely needs to be merged asap!

AndySDH commented:
Yes, this absolutely needs to be merged asap! @wilsonge

wilsonge committed:
Merge branch '4.0-dev' into fields-subfields-4.0

wilsonge commented:
Can we get a test of this done fully please. And I'm happy to get this merged before we tag the next alpha on Thursday

AndySDH commented:
Test was fully done, only minor things left to address are in my previous comment on September 8.

wilsonge commented:
OK. Then I'm happy to commit. @continga please ensure you at least sort the migration script and the calendar null date issue

richard67 commented:
@wilsonge 

> @continga please ensure you at least sort the migration script and the calendar null date issue

If you mean the null date usage in the insert statement for the extensions table in the joomla.sql files and the update sql files: This will be fixed with my PR #26491 . I've updated it so it contains the changes from this PR here and then corrects the nulldates for the checked_out_time columns to real null values like for all other extensions.

Quy commented:
Lets remove the `Render values` description. You have to read the documentation anyway to fully understand the use of subfields.

Quy commented:
@richard67 Null date check here: https://github.com/joomla/joomla-cms/pull/24711/files#diff-60b1fc5d2dc0bc1d69d7c3caa257e881R1058

richard67 commented:
@Quy Could be maybe ok because on js side, will possibily be set to the right value when writing to db in php. I will check later today or tomorrow.

Quy commented:
@continga  `Invalid argument supplied for foreach() in \libraries\src\Form\FormField.php on line 1124`

To reproduce:
- Create a subform field
- Edit an article (leave subform field empty)
- Save the article
- Check PHP error log

AndySDH commented:
@continga Woud be good if you could add an option for Repeatable Subfields to select a maximum number of rows that can be added (how many times it can be repeated).

AndySDH commented:
@continga Remember to look after any possible tweak, fixes, and the migration script. The PR has been merged like you requested, now please don't just forget about it :P

micker commented:
hello 2 questions about migration of repetable field to subfield
did you have some information about better migration ?
did you think we can have subfield for j3 to prepare a smother migration ?

AndySDH commented:
Well, a PR was ready for 3.9, but it got asked to be rebased for 4.0
See: https://github.com/joomla/joomla-cms/pull/22446
